### PR TITLE
Add Azure Active Directory authentication support

### PR DESCRIPTION
**Pull request by jsantos17**
[ch10044]

This PR adds support for Azure Active Directory authentication. This is necessary in the Avalanche destination. Will file PRs depending on this for `quasar-datasource-azure` and `quasar-destination-azure`.

### CODE REVIEW

File: azure/src/main/scala/quasar/blobstore/azure/Azure.scala [hunk 1]
```diff
@@ -17,43 +17,98 @@
...
+      Functor[F].compose[Expires].map(mkCredentials(cfg.credentials)) { creds =>
+        val pipeline =
+          StorageURL.createPipeline(
+            creds, new PipelineOptions().withLoggingOptions(new LoggingOptions(3000, true)))
+
+        new ServiceURL(url, pipeline).createContainerURL(cfg.containerName.value)
+      }
     }
+
+  def refContainerUrl[F[_]: ConcurrentEffect: Timer](cfg: Config)
```
Comment (jsantos17) [inline azure/src/main/scala/quasar/blobstore/azure/Azure.scala:90]: Unlike `SharedKey` credentials, AAD credentials expire after a set amount of time. This helper function returns a reference with the `ContainerURL` inside and a function to refresh the token. Whenver the second element of the tuple is sequenced into a computation, the `Ref` is updated with newly acquired credentials.

This is a helper function, implementation are free to use the lower level `mkContainerUrl` if they so choose.

See [AzureDestination](https://github.com/slamdata/quasar-destination-azure/blob/6d775359aa7c4ea70775d1333dd139fe655e10a8/core/src/main/scala/quasar/destination/azure/AzureDestination.scala#L40) for an example on it usage.

File: azure/src/main/scala/quasar/blobstore/azure/config.scala [hunk 2]
```diff
@@ -29,14 +33,41 @@ final case class StorageUrl(value: String)
...
+    clientSecret: ClientSecret) extends AzureCredentials
+}
 
 final case class MaxQueueSize(value: Int Refined Positive)
 
 object MaxQueueSize {
   def default: MaxQueueSize = MaxQueueSize(10)
 }
 
+final case class Expires[A](value: A, expiresAt: OffsetDateTime)
```
Comment (jsantos17) [inline azure/src/main/scala/quasar/blobstore/azure/config.scala:59]: Opinions on this? This is just a newtyped `(,) a` and its corresponding `Functor` instance.

File: azure/src/main/scala/quasar/blobstore/azure/Azure.scala [hunk 3]
```diff
@@ -17,43 +17,98 @@
...
-          new PipelineOptions().withLoggingOptions(
-            new LoggingOptions(3000, true))))
-      serviceUrl.createContainerURL(cfg.containerName.value)
+  def mkContainerUrl[F[_]](cfg: Config)(implicit F: ConcurrentEffect[F]): F[Expires[ContainerURL]] =
+    F.catchNonFatal(new URL(cfg.storageUrl.value)) flatMap { url =>
+      // Azure SDK changed its logging behavior in 10.3.0 (compared to 10.2.0).
+      // In 10.2.0 the standard `new PipelineOptions` could be used, but in 10.3.0
+      // this results in messages being logged to console in eg quasar's repl
+      // (despite having no console logger configured).
+      // exactly (e.g. `withLogger`) but adding `LoggingOptions` with
```
Comment (rintcius) [inline azure/src/main/scala/quasar/blobstore/azure/Azure.scala:74]: Looks like comment line above `exactly ...` is lost

File: azure/src/main/scala/quasar/blobstore/azure/config.scala (hunk 2 above)
Comment (rintcius) [reply azure/src/main/scala/quasar/blobstore/azure/config.scala:59]: Looks good to me!

File: azure/src/main/scala/quasar/blobstore/azure/Azure.scala (hunk 1 above)
Comment (rintcius) [reply azure/src/main/scala/quasar/blobstore/azure/Azure.scala:90]: Nice!

**Review by rintcius — COMMENTED** [review]
Looks good! - just 1 minor thing

File: azure/src/main/scala/quasar/blobstore/azure/Azure.scala (hunk 3 above)
Comment (jsantos17) [reply azure/src/main/scala/quasar/blobstore/azure/Azure.scala:74]: Oh snap. Will fix.

**Comment by rintcius** [comment]
@jsantos17 feel free to merge this PR and the 2 PRs that depend on this one, after re-adding that comment line

**Review by rintcius — APPROVED** [review]

**Comment by jsantos17** [comment]
[ch10115]

**Pull request merged**